precise/iaf_psc_exp_ps_lossless.cpp

```diff
@@ -0,0 +1,622 @@
+  {
+    const double expm1_tau_m  = numerics::expm1(-dt/P_.tau_m_);
+    
+    const double P20    = -P_.tau_m_ / P_.c_m_ * expm1_tau_m;
+    const double P21_ex = -P_.tau_m_*P_.tau_ex_ / (P_.tau_m_-P_.tau_ex_) / P_.c_m_ * (expm1_tau_ex-expm1_tau_m);
+    const double P21_in = -P_.tau_m_*P_.tau_in_ / (P_.tau_m_-P_.tau_in_) / P_.c_m_ * (expm1_tau_in-expm1_tau_m);
+    
+    S_.y2_  = P20*(P_.I_e_+S_.y0_) + P21_ex*S_.I_syn_ex_ + P21_in*S_.I_syn_in_ + expm1_tau_m*S_.y2_ + S_.y2_;
+  }
+  S_.I_syn_ex_ = S_.I_syn_ex_*expm1_tau_ex + S_.I_syn_ex_;
+  S_.I_syn_in_ = S_.I_syn_in_*expm1_tau_in + S_.I_syn_in_;
+}
+
+void nest::iaf_psc_exp_ps_lossless::emit_spike_(const Time & origin, const long lag, const double t0,  const double dt)
+{
+  // we know that the potential is subthreshold at t0, super at t0+dt
+  
+  // compute spike time relative to beginning of step
+  const double spike_offset = V_.h_ms_ - (t0 + bisectioning_(dt));
+  
+  set_spiketime(Time::step(origin.get_steps() + lag + 1));
+  S_.last_spike_offset_ = spike_offset;  
+  
+  // reset neuron and make it refractory
+  S_.y2_ = P_.U_reset_;
+  S_.is_refractory_ = true;
+  
+  // send spike
+  SpikeEvent se;
+  
+  se.set_offset(spike_offset);
+  kernel().event_delivery_manager.send(*this, se, lag);
+}
+
+void nest::iaf_psc_exp_ps_lossless::emit_instant_spike_(const Time & origin, const long lag,
+						      const double spike_offs) 
+{
+  assert( S_.y2_ >= P_.U_th_ );  // ensure we are superthreshold
+  
+  // set stamp and offset for spike
+  set_spiketime(Time::step(origin.get_steps() + lag + 1));
+  S_.last_spike_offset_ = spike_offs;
+  
+  // reset neuron and make it refractory
+  S_.y2_ = P_.U_reset_;
+  S_.is_refractory_ = true;
+  
+  // send spike
+  SpikeEvent se;
+  
+  se.set_offset(S_.last_spike_offset_);
+  kernel().event_delivery_manager.send(*this, se, lag);
+}
+
+inline double nest::iaf_psc_exp_ps_lossless::bisectioning_(const double dt) const
+{
+ 
+  double root = 0.0;
+
+  double y2_root = V_.y2_before_;
+
+  double div = 2.0;
+  while ( fabs(P_.U_th_-y2_root) > 1e-14 and (dt/div > 0.0) )
+  {
+    if ( y2_root > P_.U_th_ )
+      root -= dt/div;
+    else
+      root += dt/div;
+    
+    div *= 2.0;
+    
+    const double expm1_tau_ex = numerics::expm1(-root/P_.tau_ex_);
+    const double expm1_tau_in = numerics::expm1(-root/P_.tau_in_);
+    const double expm1_tau_m  = numerics::expm1(-root/P_.tau_m_);
+    
+    const double P20    = -P_.tau_m_ / P_.c_m_ * expm1_tau_m;
+    const double P21_ex = -P_.tau_m_*P_.tau_ex_ / (P_.tau_m_-P_.tau_ex_) / P_.c_m_ * (expm1_tau_ex-expm1_tau_m);
+    const double P21_in = -P_.tau_m_*P_.tau_in_ / (P_.tau_m_-P_.tau_in_) / P_.c_m_ * (expm1_tau_in-expm1_tau_m);
+    
+    y2_root = P20*(P_.I_e_+V_.y0_before_) + P21_ex*V_.I_syn_ex_before_ + P21_in*V_.I_syn_in_before_ + expm1_tau_m*V_.y2_before_ + V_.y2_before_;
+  }
+  return root;
+}
+
+/* Conventional spike detection algorithms propagate the initial state forwards in time 
```

Comment:
This comment should be placed in the header file as Doxygen comment and be formatted as a Doxygen comment. I feel that the comment is rather long. It should provide compact information to programmers working with the code. The overall description of the algorithm could be in the overall documentation of the neuron model and may also point to the pertaining parts/equations of the paper.